evm/utils.js

```diff
@@ -852,6 +861,73 @@ const mainProcessor = async (options, processCommand, save = true, catchErr = fa
         }
     }
 
+    if (options.parallel && chains.length > 1) {
+        const cmds = process.argv.filter((command) => command);
+        let chainCommandIndex = -1;
+        let skipPrompt = false;
+
+        for (let commandIndex = 0; commandIndex < cmds.length; commandIndex++) {
+            const cmd = cmds[commandIndex];
+
+            if (cmd === '-n' || cmd === '--chainName' || cmd === '--chainNames') {
+                chainCommandIndex = commandIndex;
+            } else if (cmd === '--parallel') {
+                cmds[commandIndex] = '--saveChainSeparately';
+            } else if (cmd === '-y' || cmd === '--yes') {
+                skipPrompt = true;
+            }
+        }
+
+        if (!skipPrompt) {
+            cmds.push('-y');
+        }
+
+        const successfullChains = [];
+
+        const executeChain = (chainName) => {
+            const chain = config.chains[chainName.toLowerCase()];
+
+            if (
+                chainsToSkip.includes(chain.name.toLowerCase()) ||
+                chain.status === 'deactive' ||
+                (chain.contracts && chain.contracts[options.contractName]?.skip)
+            ) {
+                printWarn('Skipping chain', chain.name);
+                return Promise.resolve();
+            }
+
+            return new Promise((resolve) => {
+                cmds[chainCommandIndex + 1] = chainName;
+
+                exec(cmds.join(' '), { stdio: 'inherit' }, (error, stdout) => {
+                    printInfo('-------------------------------------------------------');
+                    printInfo(`Logs for ${chainName}`, stdout);
+
+                    if (error) {
+                        printError(`Error while running script for ${chainName}`, error);
+                    } else {
+                        successfullChains.push(chainName);
+                        printInfo(`Finished running script for chain`, chainName);
+                    }
+
+                    resolve();
+                });
+            });
+        };
+
+        await Promise.all(chains.map(executeChain));
+
+        if (save) {
+            for (const chainName of successfullChains) {
+                config.chains[chainName.toLowerCase()] = loadParallelExecutionConfig(options.env, chainName);
+            }
+
+            saveConfig(config, options.env);
+        }
+
+        return;
+    }
+
```

Comment:
i don't see a need for separate code like above to handle parallel execution. Create a chainProcessor function that handles most of the `for` loop body. Depending on sequential/parallel mode, it can be run with an await or the promise can be collected, and we await at the end.

To deal with configs in parallel mode, you can give a copy of it to each call so they don't write to the same object. And then you can merge the chain specific changes at the end. No need to save configs separately.

Reply:
An initial complex solution is usually not the right one, I recommend spending more time to come up with a more elegant approach and discuss it during syncs if needed, otherwise codebase can easily become unmaintainable.